### PR TITLE
bpo-30872: Update the curses docs to Python 3.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue30872

### CODE REVIEW

File: Doc/library/curses.ascii.rst [hunk 1]
```diff
@@ -176,7 +176,7 @@ C library:
 
    Checks for a non-ASCII character (ordinal values 0x80 and above).
 
-These functions accept either integers or strings; when the argument is a
+These functions accept either integers or single-character strings; when the argument is a
```
Comment (mlouielu) [inline Doc/library/curses.ascii.rst:179]: Should this be changed? It accept multi-character strings, too. Alought it will raise `TypeError` by `ord`.

Also, after this paragraph, it has written this `Note that all these functions check ordinal bit values derived from the first character of the string you pass in`.

Comment (mlouielu) [reply Doc/library/curses.ascii.rst:179]: Does `single-character string` as same as you mention in note? (`it can be specified as a Unicode string or a byte string.`). If yes, but these function can't accept byte string:

```python
>>> curses.ascii.ascii(b'1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/curses/ascii.py", line 75, in ascii
    return _ctoi(c) & 0x7f
TypeError: unsupported operand type(s) for &: 'bytes' and 'int'
>>> curses.ascii.ctrl(b'1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.6/curses/ascii.py", line 81, in ctrl
    return _ctoi(c) & 0x1f
TypeError: unsupported operand type(s) for &: 'bytes' and 'int'
```

File: Doc/library/curses.rst [hunk 2]
```diff
@@ -219,8 +227,8 @@ The module :mod:`curses` defines the following functions:
 
 .. function:: getsyx()
 
-   Return the current coordinates of the virtual screen cursor in y and x.  If
-   leaveok is currently true, then -1,-1 is returned.
+   Return the current coordinates of the virtual screen cursor as a tuple
+   ``(y, x)``.  If **leaveok** is currently true, then ``(-1, -1)`` is returned.
```
Comment (mlouielu) [inline Doc/library/curses.rst]: true should be cast to ` ``True`` `?

File: Doc/library/curses.ascii.rst (hunk 1 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.ascii.rst:179]: Raising TypeError doesn't mean that the function accepts multi-character strings.

"string" means just string here. Although I think it is worth to add the support of bytes here, this PR changes only the documentation.

File: Doc/library/curses.rst (hunk 2 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]: Not sure.

File: Doc/library/curses.ascii.rst [hunk 3]
```diff
@@ -176,7 +176,7 @@ C library:
...
 string, it is first converted using the built-in function :func:`ord`.
 
 Note that all these functions check ordinal bit values derived from the  first
```
Comment (bitdancer) [inline Doc/library/curses.ascii.rst]: "from the first character of the string you pass in" seems confusing, given that the string may only have one character.  Perhaps change it to "from the character in the string you pass in".  (I realize this is an "existing bug", but we might as well make it more precise while we are editing the rest).

File: Doc/library/curses.rst [hunk 4]
```diff
@@ -175,17 +183,17 @@ The module :mod:`curses` defines the following functions:
...
 
 .. function:: filter()
 
    The :func:`.filter` routine, if used, must be called before :func:`initscr` is
-   called.  The effect is that, during those calls, :envvar:`LINES` is set to 1; the
-   capabilities clear, cup, cud, cud1, cuu1, cuu, vpa are disabled; and the home
-   string is set to the value of cr. The effect is that the cursor is confined to
+   called.  The effect is that, during those calls, :envvar:`LINES` is set to ``1``; the
+   capabilities **clear**, **cup**, **cud**, **cud1**, **cuu1**, **cuu**, **vpa** are disabled; and the **home**
+   string is set to the value of **cr**. The effect is that the cursor is confined to
```
Comment (bitdancer) [inline Doc/library/curses.rst]: I haven't seen two asterisks used before.  What convention does this follow?  I'm thinking these should be single `*`s?

File: Doc/library/curses.rst (hunk 2 above)
Comment (bitdancer) [reply Doc/library/curses.rst]: Probably it should be ``True``.  Even if the module isn't actually using booleans internally, *conceptually* it is; leaveok can really only have boolean values.

File: Doc/library/curses.rst [hunk 5]
```diff
@@ -313,32 +321,32 @@ The module :mod:`curses` defines the following functions:
...
    defined only after the call to :func:`initscr`.
 
 
-.. function:: meta(yes)
+.. function:: meta(flag)
 
-   If *yes* is 1, allow 8-bit characters to be input. If *yes* is 0,  allow only
-   7-bit chars.
+   If an argument is ``True``, allow 8-bit characters to be input.  If an
+   argument is ``False``,  allow only 7-bit chars.
```
Comment (bitdancer) [inline Doc/library/curses.rst]: `s/an argument/*flag*/`

Comment (vadmium) [inline Doc/library/curses.rst]: _the_ argument

File: Doc/library/curses.rst (hunk 4 above)
Comment (vadmium) [inline Doc/library/curses.rst]: Double asterisks mark up as bold don’t they? Perhaps italics (single asterisks) would be more appropriate.

File: Doc/library/curses.rst [hunk 6]
```diff
@@ -506,16 +514,17 @@ The module :mod:`curses` defines the following functions:
 
 .. function:: setsyx(y, x)
 
-   Set the virtual screen cursor to *y*, *x*. If *y* and *x* are both -1, then
-   leaveok is set.
+   Set the virtual screen cursor to *y*, *x*. If *y* and *x* are both ``-1``, then
+   **leaveok** is set.
```
Comment (bitdancer) [inline Doc/library/curses.rst]: I think this should be a link to the function, and it should probably say "set true".

File: Doc/library/curses.rst [hunk 7]
```diff
@@ -562,13 +572,13 @@ The module :mod:`curses` defines the following functions:
...
    Return the value of the string capability corresponding to the terminfo
-   capability name *capname*.  ``None`` is returned if *capname* is not a string
+   capability name *capname* as a byte string.  ``None`` is returned if *capname* is not a string
    capability, or is canceled or absent from the terminal description.
 
 
 .. function:: tparm(str[, ...])
 
-   Instantiate the string *str* with the supplied parameters, where *str* should
+   Instantiate the byte string *str* with the supplied parameters, where *str* should
```
Comment (bitdancer) [inline Doc/library/curses.rst]: tparm can't be called using keyword syntax, so we could change "str" to "capability" so we can use "bytes object" here.

File: Doc/library/curses.rst [hunk 8]
```diff
@@ -562,13 +572,13 @@ The module :mod:`curses` defines the following functions:
 .. function:: tigetstr(capname)
 
    Return the value of the string capability corresponding to the terminfo
-   capability name *capname*.  ``None`` is returned if *capname* is not a string
+   capability name *capname* as a byte string.  ``None`` is returned if *capname* is not a string
    capability, or is canceled or absent from the terminal description.
```
Comment (bitdancer) [inline Doc/library/curses.rst]: I think we should rewrite this as `...*capname* as a bytes object.  ``None is returned if *capname* is not a terminfo "string capability", or is...`

File: Doc/library/curses.rst [hunk 9]
```diff
@@ -629,7 +639,7 @@ The module :mod:`curses` defines the following functions:
 .. function:: use_env(flag)
 
    If used, this function should be called before :func:`initscr` or newterm are
-   called.  When *flag* is ``False``, the values of lines and columns specified in the
+   called.  When an argument is ``False``, the values of lines and columns specified in the
```
Comment (bitdancer) [inline Doc/library/curses.rst]: `s/an argument/*flag*/`

File: Doc/library/curses.rst [hunk 10]
```diff
@@ -783,7 +786,7 @@ the following methods and attributes:
             window.chgat(y, x, num, attr)
 
    Set the attributes of *num* characters at the current cursor position, or at
-   position ``(y, x)`` if supplied. If no value of *num* is given or *num* = -1,
+   position ``(y, x)`` if supplied. If no value of *num* is given or *num* = ``-1``,
```
Comment (bitdancer) [inline Doc/library/curses.rst]: I would change this to `If no value is given for *num*, or ``-1`` is given,`

File: Doc/library/curses.rst [hunk 11]
```diff
@@ -796,9 +799,9 @@ the following methods and attributes:
    call to :meth:`refresh`.
 
 
-.. method:: window.clearok(yes)
+.. method:: window.clearok(flag)
 
-   If *yes* is 1, the next call to :meth:`refresh` will clear the window
+   If an argument is ``True``, the next call to :meth:`refresh` will clear the window
```
Comment (bitdancer) [inline Doc/library/curses.rst]: `s/an argument/*flag*`

File: Doc/library/curses.rst [hunk 12]
```diff
@@ -881,14 +884,15 @@ the following methods and attributes:
...
-   no-delay mode, -1 is returned if there is no input, else :func:`getch` waits
+   no-delay mode, ``-1`` is returned if there is no input, else the function waits
    until a key is pressed.
 
 
 .. method:: window.get_wch([y, x])
 
    Get a wide character. Return a character for most keys, or an integer for
    function keys, keypad keys, and other special keys.
+   In no-delay mode, an exception is raised if there is no input.
```
Comment (bitdancer) [inline Doc/library/curses.rst]: Again, stick with imperative: "raise an exception if"

File: Doc/library/curses.rst [hunk 13]
```diff
@@ -881,14 +884,15 @@ the following methods and attributes:
 
    Get a character. Note that the integer returned does *not* have to be in ASCII
    range: function keys, keypad keys and so on return numbers higher than 256. In
-   no-delay mode, -1 is returned if there is no input, else :func:`getch` waits
+   no-delay mode, ``-1`` is returned if there is no input, else the function waits
```
Comment (bitdancer) [inline Doc/library/curses.rst]: We should fix the tense here. The first part of the paragraph is in imperative mode, so this should say `In no-delay mode, return ``-1`` if there is no input, otherwise wait until a key is pressed`.  I might also change "return numbers higher than 256" to "are represented by numbers higher than 265".

File: Doc/library/curses.rst [hunk 14]
```diff
@@ -933,21 +940,21 @@ the following methods and attributes:
 
 .. method:: window.idcok(flag)
 
-   If *flag* is ``False``, curses no longer considers using the hardware insert/delete
-   character feature of the terminal; if *flag* is ``True``, use of character insertion
+   If an argument is ``False``, curses no longer considers using the hardware insert/delete
+   character feature of the terminal; if an argument is ``True``, use of character insertion
```
Comment (bitdancer) [inline Doc/library/curses.rst]: `*flag*` was correct here, this should be reverted.

File: Doc/library/curses.rst [hunk 15]
```diff
@@ -933,21 +940,21 @@ the following methods and attributes:
...
    and deletion is enabled.  When curses is first initialized, use of character
    insert/delete is enabled by default.
 
 
-.. method:: window.idlok(yes)
+.. method:: window.idlok(flag)
 
-   If called with *yes* equal to 1, :mod:`curses` will try and use hardware line
+   If called with an argument equal to ``True``, :mod:`curses` will try and use hardware line
```
Comment (bitdancer) [inline Doc/library/curses.rst]: `s/an argument/*flag*/`

File: Doc/library/curses.rst [hunk 16]
```diff
@@ -933,21 +940,21 @@ the following methods and attributes:
...
    editing facilities. Otherwise, line insertion/deletion are disabled.
 
 
 .. method:: window.immedok(flag)
 
-   If *flag* is ``True``, any change in the window image automatically causes the
+   If an argument is ``True``, any change in the window image automatically causes the
```
Comment (bitdancer) [inline Doc/library/curses.rst]: revert this

File: Doc/library/curses.rst [hunk 17]
```diff
@@ -1022,20 +1029,20 @@ the following methods and attributes:
    :meth:`refresh`; otherwise return ``False``.
 
 
-.. method:: window.keypad(yes)
+.. method:: window.keypad(flag)
 
-   If *yes* is 1, escape sequences generated by some keys (keypad,  function keys)
-   will be interpreted by :mod:`curses`. If *yes* is 0, escape sequences will be
+   If an argument is ``True``, escape sequences generated by some keys (keypad,  function keys)
+   will be interpreted by :mod:`curses`. If an argument is ``False``, escape sequences will be
```
Comment (bitdancer) [inline Doc/library/curses.rst]: `s/an argument/*flag*/`

File: Doc/library/curses.rst [hunk 18]
```diff
@@ -1022,20 +1029,20 @@ the following methods and attributes:
...
    left as is in the input stream.
 
 
-.. method:: window.leaveok(yes)
+.. method:: window.leaveok(flag)
 
-   If *yes* is 1, cursor is left where it is on update, instead of being at "cursor
+   If an argument is ``True``, cursor is left where it is on update, instead of being at "cursor
```
Comment (bitdancer) [inline Doc/library/curses.rst]: `s/an argument/*flag*/`

Ok, I'm going to stop noting this every time, you get the idea :)

File: Doc/library/curses.rst [hunk 19]
```diff
@@ -1202,7 +1209,7 @@ the following methods and attributes:
 
 .. method:: window.syncok(flag)
 
-   If called with *flag* set to ``True``, then :meth:`syncup` is called automatically
+   If called with ``True``, then :meth:`syncup` is called automatically
```
Comment (bitdancer) [inline Doc/library/curses.rst]: `If *flag* is ``True``...`

**Review by bitdancer — CHANGES_REQUESTED** [review]
I checked, and we use "bytes object" 98 times in library docs, and "byte string" 42 times. So I think we are shifting to using "bytes object".  I would like to see all the instances of "byte string" changed to "bytes object" here.

**Comment by serhiy-storchaka** [comment]
I intentionally used the term "byte string" for accenting that a bytes object is not just arbitrary couple of bits as in binary representation of floats or images, but mostly human-readable string. Do you think this makes sense?

File: Doc/library/curses.ascii.rst (hunk 3 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.ascii.rst]: Agree. Thank you for the suggestion.

The rest of the paragraph looks outdated too. Most functions was moved from the `string` module and become `str` methods. And they don't know about the character encoding because handle Unicode.

File: Doc/library/curses.rst (hunk 4 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]: I wanted to distinguish them from other names. These names are not the names of arguments (which are rendered as italic). They are the names of terminal capabilities. If you prefer, I can change them to be rendered as fixed-width (\`\`clear\`\`), but not as italic.

File: Doc/library/curses.rst (hunk 2 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]: **leaveok** is a name of the window option. It is changed by the method `window.leaveok()`. How better to reword this?

File: Doc/library/curses.rst (hunk 5 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]: I unified all boolean argument names (now all are "flag"), but removed names from descriptions because actually the name is arbitrary. Since all these functions take a single argument, there is no need to name it.

With these reasons what is better, "`the argument`" or "`*flag*`"?

File: Doc/library/curses.rst (hunk 6 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]:
```
If *y* and *x* are both ``-1``, then :meth:`~window.leaveok` set true.
```
Right?

File: Doc/library/curses.rst (hunk 7 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]: I afraid this will cause an ambiguity between capability name and value. The first argument of `tparm()` is not a capability name. It is usually a result of `tigetstr()`.

File: Doc/library/curses.rst (hunk 10 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]: Done.

File: Doc/library/curses.rst (hunk 13 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]: Done here and in other places.

**Comment by bitdancer** [comment]
I still prefer bytes object, but in this context it is not a strong preference.  The "character string" idiom is also somewhat problematic, but I don't have a better solution to suggest.

File: Doc/library/curses.rst (hunk 4 above)
Comment (bitdancer) [reply Doc/library/curses.rst]: Yes, I think that would be better.

File: Doc/library/curses.rst (hunk 5 above)
Comment (bitdancer) [reply Doc/library/curses.rst]: ``*flag*``.  Our documentation convention is that we refer to the arguments by the name given in the prototype, regardless of whether that is what is actually used in the code or not.  It would be valid English to say "the argument", but I think consistency with the bulk of the docs is a good idea.

File: Doc/library/curses.rst (hunk 6 above)
Comment (bitdancer) [reply Doc/library/curses.rst]: "is set true".  Sorry I didn't make that clear.

File: Doc/library/curses.rst (hunk 7 above)
Comment (bitdancer) [reply Doc/library/curses.rst]: Hmm.  I guess you are right.  The termcap docs are a bit ambiguous in how they use the term 'capability'  Since 'str' is in the tigetstr name, if you are sticking with bytes string it makes sense to keep it.

File: Doc/library/curses.rst [hunk 20]
```diff
@@ -786,7 +787,7 @@ the following methods and attributes:
             window.chgat(y, x, num, attr)
 
    Set the attributes of *num* characters at the current cursor position, or at
-   position ``(y, x)`` if supplied. If no value of *num* is given or *num* = ``-1``,
+   position ``(y, x)`` if supplied. If no value of *num* is given or ``-1`` is given,
```
Comment (vadmium) [inline Doc/library/curses.rst]: Why this change? I think the original read better. If you want to use words rather than code, you could write “or _num_ is −1”, or “_num_ is negative one”. Or perhaps “If _num_ is not given or is −1”.

File: Doc/library/curses.rst [hunk 21]
```diff
@@ -883,16 +884,16 @@ the following methods and attributes:
 .. method:: window.getch([y, x])
 
    Get a character. Note that the integer returned does *not* have to be in ASCII
-   range: function keys, keypad keys and so on return numbers higher than 256. In
-   no-delay mode, ``-1`` is returned if there is no input, else the function waits
-   until a key is pressed.
+   range: function keys, keypad keys and so are represented by numbers higher
```
Comment (vadmium) [inline Doc/library/curses.rst]: I would prefer to restore the word _on_. _And so on_ means _and similar cases_.

File: Doc/library/curses.rst (hunk 20 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]: It was changed on the request by @bitdancer.

File: Doc/library/curses.rst (hunk 21 above)
Comment (serhiy-storchaka) [reply Doc/library/curses.rst]: Good catch! I removed it by mistake when edited following words.

File: Doc/library/curses.rst (hunk 20 above)
Comment (bitdancer) [reply Doc/library/curses.rst]: I don't particularly remember this, but I suspect my logic was that we don't usually write "arg = <something>" when describing argument values, since that looks like code and thus is potentially confusing in this context.  I like the "if num is not given or is -1" phrasing.  Oddly enough, I think it might read better in the form "if num is -1 or is not given", but either one is fine with me.

Comment (vadmium) [reply Doc/library/curses.rst]: Either is fine by me

**Comment by serhiy-storchaka** [comment]
Should I do anything more @bitdancer?

**Comment by serhiy-storchaka** [comment]
Ping.

**Comment by serhiy-storchaka** [comment]
@Mariatta, running `cherry_picker` manually didn't expose any conflicts. I don't know why @miss-islington failed with backporting.

**Comment by Mariatta** [comment]
Thanks @serhiy-storchaka . It's strange; quickly reading the log, there's an exception when it was trying to do `git fetch upstream`..
However I haven't find out the cause of that..

**Comment by bitdancer** [comment]
I don't know if it is relevant, but in the past few days on the commercial project I'm working on we've seen an increased incidence of codeship CI jobs failing because the pull from github fails, and then the restarted job will work fine.  That is, this may be a github reliability issue.

**Comment by Mariatta** [comment]
Thanks @bitdancer :) Yeah I just restarted miss-islington hoping to get it in clean state ..

I think the next backport should work..
